presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownAggregation.java

```diff
@@ -0,0 +1,118 @@
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule.Context;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class DereferencePushDownAggregation
+        extends DereferencePushDownRule<AggregationNode>
+{
+    public DereferencePushDownAggregation(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, aggregation());
+    }
+
+    @Override
+    protected Result pushDownDereferences(Context context, AggregationNode aggregationNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = aggregationNode.getOutputSymbols();
+        Map<Symbol, Expression> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> !isGroupingKey(entry.getKey(), aggregationNode))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+        if (pushdownExpressions.isEmpty()) {
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
+        for (Aggregation aggregation : aggregationNode.getAggregations().values()) {
+            expressionBuilder.addAll(aggregation.getCall().getArguments());
+        }
+        List<DereferenceExpression> aggregations = expressionBuilder.build().stream()
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .collect(toImmutableList());
+
+        Map<Symbol, DereferenceExpression> aggregationDereferences = aggregations.stream()
+                .distinct()
+                .filter(expression -> !expressions.containsKey(expression))
+                .collect(toImmutableMap(expression -> newSymbol(expression, context, metadata, sqlParser), Function.identity()));
```

Comment:
These transformations can be combined. Intermediate variables expressionBuilder and aggregations can be removed.

```
        Map<Symbol, DereferenceExpression> aggregationDereferences = aggregationNode.getAggregations().values().stream()
                .flatMap(aggregation -> aggregation.getCall().getArguments().stream())
                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
                .distinct()
                .filter(expression -> !expressions.containsKey(expression))
                .collect(toImmutableMap(expression -> newSymbol(expression, context, metadata, sqlParser), Function.identity()));
```